File: 150-problems/002-is-anagram.go
package problems

func Is_anagram(s, t string) bool {
	if len(s) != len(t) {
		return false
	}
	checkMap := make(map[string]int)

	for i := range s {
		sc := string(s[i])
		checkMap[sc] += 1
		tc := string(t[i])
		checkMap[tc] -= 1
	}

	for _,v := range checkMap {
		if v != 0 {
			return false
		}
	}

	return true
}

File: 150-problems/002-is-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # If the length of the strings are not equal, then they are not anagrams
        if len(s) != len(t):
            return False
        # Create a list of 26 elements, each element representing a character in the alphabet
        chrs = [0]*26
        # Iterate through the strings and increment the value of the character in the list
        for i in range(len(t)):
            # add the chars in s and subtract the chars in t from the list
            chrs[ord(s[i])-ord('a')] += 1
            chrs[ord(t[i])-ord('a')] -= 1
        
        # If the list contains any non-zero values, then the strings are not anagrams
        for n in chrs:
            if n != 0:
                return False
        
        return True

# Test cases
s = Solution()
print(s.isAnagram("anagram", "nagaram")) # True
print(s.isAnagram("rat", "car")) # False
print(s.isAnagram("a", "ab")) # False
print(s.isAnagram("a", "a")) # True